Update to 1.19.3, add event-based registry inspection and modification API

AlexTMjugador opened the pull request:
A Minecraft version between 1.19.1 and 1.19.3 changed several internal implementation details related to registries and the network codec that broke compatibility with this plugin:

![Plugin load exception on 1.19.3](https://user-images.githubusercontent.com/7822554/213530528-35fe653f-55c8-417f-b8b5-3d6594dcc855.png)

I'm interested in getting this plugin to work on a Paper 1.19.3 server, so I dug into the Minecraft class changes and updated the plugin accordingly. I've tested the changes lightly on a local server by using the example override file mentioned in the README.

AlexTMjugador committed:
Update to 1.19.3

Owen1212055 commented on nms/src/main/java/com/owen1212055/biomevisuals/nms/hooks/RegistryHook.java:
```diff
@@ -18,22 +18,22 @@ public static void injectCodec(Map<HookType, List<KeyedOverride>> hooks, Logger
 
             @Override
             public <T> DataResult<T> encode(RegistryAccess input, DynamicOps<T> ops, T prefix) {
-                DataResult<JsonElement> result = CAPTURED_CODEC.encode(input, JsonOps.INSTANCE, JsonOps.INSTANCE.empty());
```
Please specify the types rather than using var.

AlexTMjugador replied:
Done. Thanks for the prompt review! :heart:

Owen1212055 reviewed (changes requested):
Little thing.

Thank you so much for your work! Thanks for using this plugin too. 😄

Owen1212055 commented:
And if you wanna give it a shot, id recommend making registry injection an event rather than having to register it each time.

We can still allow manual registration, but an event in addition would be cool. If you don't want to no worries.

AlexTMjugador committed:
Do not infer several variable types at NMS RegistryHook

AlexTMjugador commented:
I would not mind giving that refactor some attention, but I'm not sure how you envision events altering the lifecycle of registry injections, so I don't feel confident doing it yet. Would you mind elaborating on what you want to see? :heart:

Owen1212055 commented:
> I would not mind giving that refactor some attention, but I'm not sure how you envision events altering the lifecycle of registry injections, so I don't feel confident doing it yet. Would you mind elaborating on what you want to see? :heart:

I imagine a sort of RegistryOverrideEvent which will be triggered when the codec runs it's serialization logic. The current "injections" which are provided by the api are a bit limiting, and in general this event would offer as a replacement for those who want more powerful behavior. 

Internally, there will need to be something that listens to this event and uses it for the already existing override system. If that makes sense.

AlexTMjugador commented:
Thanks for the explanation! It helped me to understand your vision more clearly. I'll give your idea a shot and let you know when I give up or have any results to share :smile:

AlexTMjugador committed:
Refactor registry hook logic to be event-based

The hooked codec now fires a custom RegistrySendEvent that is used
internally to implement biome overrides. In addition, this event can be
listened to by third-party plugins to do further advanced, low-level
registry manipulation and inspection.

AlexTMjugador committed:
Set HOOK_ACTIVE flag as intended

This flag previously was never set, causing the /biomevisuals overrides
command to always report that the hooks were not active.

AlexTMjugador commented:
I did it! :tada: 

The latest commits add a new `RegistrySendEvent` to the API that is fired by the hooked codec with the registry data that is about to be sent. Listeners can inspect that data and modify it. Any modifications will be applied to the data sent to players unless the event is cancelled. The plugin is now a listener of this event, and uses it to implement the override logic.

Owen1212055 commented on nms/src/main/java/com/owen1212055/biomevisuals/nms/hooks/RegistryHook.java:
```diff
@@ -34,32 +36,37 @@ public <T> DataResult<T> encode(RegistryAccess input, DynamicOps<T> ops, T prefi
+                    Map<NamespacedKey, JsonObject> registryEntryHolders = new HashMap<>(registry.size());
                     for (JsonElement jsonElement : registry) {
-                        registeredObjects.put(jsonElement.getAsJsonObject().get("name").getAsString(), jsonElement.getAsJsonObject().getAsJsonObject("element"));
+                        var namespacedKey = Objects.requireNonNull(NamespacedKey.fromString(jsonElement.getAsJsonObject().get("name").getAsString()));
```
var + don't use Objects.requireNonNull.
In general your IDE probably screams at you to add this, but, it's not needed.

Owen1212055 commented on nms/src/main/java/com/owen1212055/biomevisuals/nms/hooks/RegistryHook.java:
```diff
@@ -34,32 +36,37 @@ public <T> DataResult<T> encode(RegistryAccess input, DynamicOps<T> ops, T prefi
-                            continue;
-                        }
+                    // Let event handlers observe and modify the registry data.
+                    final var sendEvent = new RegistrySendEvent(hookedTypeKey, registryEntries);
```
var

Owen1212055 commented on nms/src/main/java/com/owen1212055/biomevisuals/nms/hooks/RegistryHook.java:
```diff
@@ -34,32 +36,37 @@ public <T> DataResult<T> encode(RegistryAccess input, DynamicOps<T> ops, T prefi
+                        continue;
+                    }
+
+                    for (final var overriddenEntry : registryEntries.entrySet()) {
```
var

Owen1212055 commented on plugin/src/main/java/com/owen1212055/biomevisuals/Main.java:
```diff
@@ -92,6 +98,25 @@ public void onDisable() {
+            return;
+        }
+
+        for (final var keyedOverride : keyedOverrides) {
```
var

Owen1212055 reviewed (changes requested):
This is really good!!!
This whole concept is much cleaner in general and will hopefully prove the api more powerful.

One thing that i'm a little bit iffy on is that the underlying JsonObject is exposed. See how we have stuff like BiomeData, it would be nice to instead pass that instead of the raw json.
Generally you're moving the jsonobject "merging" from the internal implementation to the server. 

Perhaps we could have something in the API like a  RegistryKey\<BiomeData\> and RegistryKey\<DimensionData\> (unused).

I donno know, generics with evens like this are a bit iffy, but try playing around.
Overall really happy with this, nice work. 😄

AlexTMjugador committed:
Address var usage as per review comments

AlexTMjugador committed:
Comment out the dimension hook type at HookType. Micro-optimize values map

AlexTMjugador committed:
Replace event map entry null check in hooked codec by unmodifiable map

Because unmodified maps can't have entries added or removed (but they
can be modified), null checks are not necessary. This should be slightly
more performant in common cases, and alert users of wrong code promptly.

AlexTMjugador committed:
Fix typo

Owen1212055 commented on plugin/src/main/java/com/owen1212055/biomevisuals/Main.java:
```diff
@@ -105,7 +105,7 @@ public void onRegistrySend(final @NotNull RegistrySendEvent event) {
         }
 
-        for (final var keyedOverride : keyedOverrides) {
+        for (var keyedOverride : keyedOverrides) {
```
You didn't actually remove the vars. 😢
Just replace it with the actual types.

AlexTMjugador replied:
My bad, I thought you wanted me to remove the `final` qualifier only. I'll get rid of `var` now :smile:

AlexTMjugador committed:
Get rid of some var. Rename HookType to RegistryType and expose in API

Exposing RegistryType to the public is necessary to make the event
payload more strongly typed.

AlexTMjugador committed:
Refactor registry send events to use strongly typed objects

Several breaking changes to the API were made, and lots of code has to
be changed to account for deserializing and serializing registry data
from and to JSON.

This commit is a continuation of the previous commit.

AlexTMjugador commented:
It has been more complicated than I anticipated, but I made the biome registry send event expose strongly-typed `BiomeData` objects instead. The code is designed in such a way that future registry type hooks (i.e., for dimension registries) should get their own registry send event class, so there are no generic type parameters in the event class itself, which I think provides the most comfortable API for end-users.

These strongly-typed objects are now deserialized too, so the `JsonAdapter` class has been moved and expanded. In addition, most registry data holder classes in the API were refactored to records, and their data model streamlined.

This is how using the new event handler API looks like after all this work. I think it's pretty nice, isn't it? :heart:

https://github.com/Owen1212055/BiomeVisuals/blob/99bc84e483af43a261d23eec27d36c7792e803b8/plugin/src/main/java/com/owen1212055/biomevisuals/Main.java#L101-L125

Owen1212055 commented on nms/src/main/java/com/owen1212055/biomevisuals/nms/ApiEntityConverter.java:
```diff
@@ -1,34 +1,91 @@
+
+        try {
+            // FIXME the particle data object is not serialized as such, so we can't deserialize it back
+            var probabilityField = AmbientParticleSettings.class.getDeclaredField("c");
```
Can you put the Mojang name next to it? And var :(

AlexTMjugador replied:
Of course, done!

Owen1212055 commented:
100%! Will look over the weekend, I recommend joining my discord (check the readme)

AlexTMjugador committed:
Address review comment

Owen1212055 committed:
Move events, add editorconfig

Owen1212055 committed:
Cleanups + general moving around

AlexTMjugador committed:
Encapsulate registry data map usage on RegistrySendEvent

AlexTMjugador committed:
Revert some biome effect data holder classes being refactored to records

AlexTMjugador committed:
Make AmbientParticle constructor private

This provides better encapsulation, in line with how other constructors
for biome effect data holder classes are also private.
